### PR TITLE
Simple nxdk audio sample

### PR DESCRIPTION
**Pull request by mborgerson**
We should have *some* demo to show how to output basic audio with nxdk. If a ~~better~~ more portable/conventional API (e.g. SDL2) is available in the future, ideally one actually providing a convenient interface to the APU, then we can replace this. Until then, this sample provides a nice and simple means to get some audio playback.

Audio comes courtesy my dog, Murdock. CC0

### CODE REVIEW

File: samples/xaudio/Makefile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+XBE_TITLE=hello
```
Comment (JayFoxRox) [inline samples/xaudio/Makefile]: Wrong title

File: samples/xaudio/main.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    current_buf = (current_buf+1) % NUM_BUFFERS;
+    callback_count++;
+}
+
+int main(void)
+{
+    XVideoSetMode(640, 480, 32, REFRESH_DEFAULT);
+
+    for (int i = 0; i < NUM_BUFFERS; i++) {
+        buffers[i] = MmAllocateContiguousMemoryEx(buffer_size,0,0x3ffb000,0,0x404);
```
Comment (JayFoxRox) [inline samples/xaudio/main.c]: Use `PAGE_READWRITE | PAGE_WRITECOMBINE` (or something?) instead of the 0x404.

Comment (JayFoxRox) [inline samples/xaudio/main.c]: nit: We use something like `#define MAX_RAM 0x3ffb000` for another sample.

File: samples/xaudio/main.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+unsigned char *voice_data;
+unsigned int voice_len;
+unsigned int voice_pos;
+
+#define NUM_BUFFERS 2
+const unsigned short buffer_size = 48*1024;
+unsigned char *buffers[NUM_BUFFERS];
+unsigned int current_buf;
+
+static void provide_samples_callback(void *pac97Device, void *data)
```
Comment (JayFoxRox) [inline samples/xaudio/main.c]: How is this called? I could imagine that this doesn't allow float as it is an interrupt (do DPCs allow float?)

File: samples/xaudio/main.c (hunk 2 above)
Comment (mborgerson) [reply samples/xaudio/main.c]: All of our samples allocate memory this way, but I agree we shouldn't use these constants

Comment (mborgerson) [reply samples/xaudio/main.c]: Same for your other comment for this line above

File: samples/xaudio/Makefile (hunk 1 above)
Comment (mborgerson) [reply samples/xaudio/Makefile]: Copy-paste bug, will fix

File: samples/xaudio/main.c (hunk 3 above)
Comment (mborgerson) [reply samples/xaudio/main.c]: It's called from a DPC. No floats are used here, but if someone uses floats here they will need to save and restore the FPU state. I added a comment to warn people.

File: samples/xaudio/Makefile (hunk 1 above)
Comment (mborgerson) [reply samples/xaudio/Makefile]: Done

File: samples/xaudio/main.c (hunk 2 above)
Comment (mborgerson) [reply samples/xaudio/main.c]: Done

Comment (mborgerson) [reply samples/xaudio/main.c]: Done with a FIXME to move this to another file. There's no common spot for it yet. We should refactor all samples that use this constant similarly.

**Comment by JayFoxRox** [comment]
For consistency, please use prefix style for commit title (`samples: ` in this case).

I have skimmed over the code earlier. I'll probably try running the sample and also listen to the audio file before giving my final okay.

**Comment by mborgerson** [comment]
> For consistency, please use prefix style for commit title (`samples: ` in this case).

Ok

**Comment by JayFoxRox** [comment]
I have skimmed over the code and tested the sample.

Merged.

---

Some final notes about things which I would like to see improved in the future:

- nxdk.wav is a bad filename, woof.wav, dog.wav, bark.wav or something would have been more suitable.
- Another audio file which is more recognizable would have been nice, ideally human voice or a simple melody (to show pitch correctness, volume and playback speed).
- Direct loading of the WAV would have been nice, but I can also understand why it wasn't done.
- The callback `data` argument is a bit confusing. It could be documented that it comes from the callback setter.
- The callback ~`device` parameter is a bit confusing, and it being `NULL` in the initial samples might lead to bugs. There could have been a second function without arguments to avoid this problem.
- The `debugPrint` cursor position could have been controlled, so the counter could be updated more often (as it currently skips a couple of numbers).
- As already suggested in the PR description, we should try to get rid of this API, so we should replace it with an SDL audio or OpenAL-like (see #20) sample instead.
- "Dead code" in Makefile for format conversion could have had a comment.
- Buffer size in `unsigned short` is very tight.
- Buffer size follows no logical formula (is `48*1024` but could have been `12288 * NUM_CHANNELS * sizeof(int16_t) // ~250ms of audio (12288 samples at 48kHz)`).

**Comment by mborgerson** [comment]
> nxdk.wav is a bad filename, woof.wav, dog.wav, bark.wav or something would have been more suitable.

I'm not sure that it is a "bad" filename. It's a placeholder file, so nxdk.wav is acceptable in my opinion. People are expected to change it.

> Another audio file which is more recognizable would have been nice, ideally human voice or a simple melody (to show pitch correctness, volume and playback speed).

That would be fine, as long as the file remains relatively small. If you do this, please find or create something that is short and CC0 for redistribution.

> Direct loading of the WAV would have been nice, but I can also understand why it wasn't done.

I'm not sure that's necessary for this small sample, but would like to see someone add WAV loader. I consider that to be outside the scope of this PR.

> The callback data argument is a bit confusing. It could be documented that it comes from the callback setter. The callback ~device parameter is a bit confusing, and it being NULL in the initial samples might lead to bugs. There could have been a second function without arguments to avoid this problem.

The prototype for the callback is copy-pasted from the xaudio API. It's good to stick with the names used by the API as I have done here. A follow-up PR could rename the parameters throughout the code base if you can think of better names.

Consumers of this particular sample code should be expected to understand the nature of a callback function and its typical arguments like application-supplied pointer and device pointer, and refer to the xaudio API if they have questions (which could be better documented, to be fair).

> The debugPrint cursor position could have been controlled, so the counter could be updated more often (as it currently skips a couple of numbers).

That's not a bad idea, might be easier to read too. Come to think of it, I did exactly that for my timer sample.

> "Dead code" in Makefile for format conversion could have had a comment.

I think it's pretty clear, but it wouldn't hurt I suppose.

> Buffer size in unsigned short is very tight.

Buffer size is defined by the API as an `unsigned short`, that's why it's used here. Using a wider type could lead to bugs if the API is not updated. Though, what is passed into the ac97 descriptor is an unsigned short of number of samples so this could actually support higher buffer sizes, but we need to change the API. I've left that for future work. Also, the 48 KiB buffer is plenty in this case--one can get away with much, much smaller buffers, at the cost of more frequent interrupts.

> Buffer size follows no logical formula (is 48*1024 but could have been 12288 * NUM_CHANNELS * sizeof(int16_t) // ~250ms of audio (12288 samples at 48kHz)).

Buffer size was selected to be "sufficiently large." If you want to PR a formula to calculate buffer size based on duration of audio content, that would be nice to have.

**Pull request merged**